sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -464,6 +465,24 @@ object OptimizeIn extends Rule[LogicalPlan] {
 }
 
 /**
+  * Convert an expression into its conjunctive normal form (CNF), i.e. AND of ORs.
+  * For example, a && b || c is normalized to (a || c) && (b || c) by this method.
+  *
+  * Refer to https://en.wikipedia.org/wiki/Conjunctive_normal_form for more information
+  */
+object CNFNormalization extends Rule[LogicalPlan] {
```

Comment:
@marmbrus @nongli do we want to do this for all expressions? If we do, maybe we should have a feature flag for this?